File: apps/api/src/services/rightnow_service.py
"""RightNow Service - Prioritizes Canvas assignments by due date urgency."""
from typing import Optional
from datetime import datetime, timedelta
from sqlalchemy.orm import Session
from ..models.db_models import StudyTask, ScheduleEvent
from ..models.schemas import RightNowResponse


def get_right_now_suggestion(db: Session, user_id: str) -> RightNowResponse:
    """
    Get the most urgent task or upcoming class suggestion.
    Prioritizes Canvas assignments by due date (nearest first).
    """
    now = datetime.utcnow()
    
    # Get all user tasks, ordered by due date
    all_tasks = db.query(StudyTask).filter(
        StudyTask.user_id == user_id
    ).order_by(StudyTask.due_at.asc()).all()
    
    # Filter to upcoming tasks (due_at >= now)
    upcoming_tasks = []
    for task in all_tasks:
        try:
            due_date = datetime.fromisoformat(task.due_at.replace('Z', '+00:00'))
            if due_date.tzinfo is None:
                due_date = datetime.fromisoformat(task.due_at)
            
            # Normalize to UTC for comparison
            if due_date.tzinfo:
                due_date = due_date.replace(tzinfo=None) - timedelta(hours=8)  # Convert SGT to UTC if needed
            
            if due_date >= now:
                upcoming_tasks.append((task, due_date))
        except (ValueError, AttributeError):
            # Skip tasks with invalid date format
            continue
    
    # Sort by urgency (nearest due first)
    if upcoming_tasks:
        upcoming_tasks.sort(key=lambda x: x[1])
        most_urgent_task, due_date = upcoming_tasks[0]
        
        # Calculate time until due
        time_until_due = due_date - now
        hours_until = time_until_due.total_seconds() / 3600
        
        # Format message based on urgency
        if hours_until < 2:
            urgency = "CRITICAL"
        elif hours_until < 24:
            urgency = "URGENT"
        elif hours_until < 48:
            urgency = "HIGH"
        else:
            urgency = "NORMAL"
        
        if hours_until < 1:
            time_str = f"{int(time_until_due.total_seconds() / 60)} minutes"
        elif hours_until < 24:
            time_str = f"{int(hours_until)} hours"
        else:
            days = int(hours_until / 24)
            time_str = f"{days} days"
        
        message = f"Focus on {most_urgent_task.course} - {most_urgent_task.title} | Due in {time_str} ({urgency})"
        
        return RightNowResponse(
            type="TASK",
            message=message,
            data={
                "task": most_urgent_task.to_dict(),
                "hoursUntil": round(hours_until, 2),
                "urgency": urgency
            }
        )
    
    # No urgent tasks - check if class starting soon (within 60 min)
    all_events = db.query(ScheduleEvent).filter(
        ScheduleEvent.user_id == user_id
    ).order_by(ScheduleEvent.day, ScheduleEvent.start_time).all()
    
    # Get current day of week
    current_day = now.strftime("%A")
    current_time_str = now.strftime("%H%M")
    current_time_int = int(current_time_str)
    
    # Check events for today or next few days
    day_order = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
    current_day_idx = day_order.index(current_day) if current_day in day_order else 0
    
    for event in all_events:
        event_day_idx = day_order.index(event.day) if event.day in day_order else 99
        
        # Check if event is today or in the future
        days_until_event = (event_day_idx - current_day_idx) % 7
        if days_until_event == 0:  # Today
            try:
                event_start_int = int(event.start_time)
                time_diff = event_start_int - current_time_int
                # Convert to minutes
                hours = time_diff // 100
                minutes = time_diff % 100
                minutes_until = hours * 60 + minutes
                
                if 0 <= minutes_until <= 60:
                    message = f"You have {event.module} {event.type} starting at {event.start_time[:2]}:{event.start_time[2:]} in {minutes_until} minutes"
                    return RightNowResponse(
                        type="CLASS",
                        message=message,
                        data={
                            "event": event.to_dict(),
                            "minutesUntil": minutes_until
                        }
                    )
            except (ValueError, AttributeError):
                continue
        elif days_until_event > 0 and days_until_event <= 1:  # Tomorrow or next day
            try:
                event_start_int = int(event.start_time)
                if event_start_int <= 1400:  # Before 2 PM
                    message = f"You have {event.module} {event.type} tomorrow at {event.start_time[:2]}:{event.start_time[2:]}"
                    return RightNowResponse(
                        type="CLASS",
                        message=message,
                        data={
                            "event": event.to_dict(),
                            "daysUntil": days_until_event
                        }
                    )
            except (ValueError, AttributeError):
                continue
    
    # No urgent tasks or upcoming classes
    return RightNowResponse(
        type="IDLE",
        message="No urgent tasks. You can work ahead!"
    )

File: apps/web/app/api/src/services/rightnowService.ts
// RightNow Service - Prioritizes Canvas assignments by due date urgency
import { StudyTask, ScheduleEvent, RightNowResponse } from '../types';
import { getUserTasks, getUserSchedule } from './dataService';

/**
 * Get the most urgent task or upcoming class suggestion.
 * Prioritizes Canvas assignments by due date (nearest first).
 */
export async function getRightNowSuggestion(userId: string): Promise<RightNowResponse> {
  const now = new Date();
  
  // Get all user tasks, ordered by due date
  const allTasks = await getUserTasks(userId, 100); // Get more tasks to filter
  
  // Filter to upcoming tasks (dueAt >= now)
  const upcomingTasks: Array<{ task: StudyTask; dueDate: Date }> = [];
  
  for (const task of allTasks) {
    try {
      const dueDate = new Date(task.dueAt);
      if (dueDate >= now) {
        upcomingTasks.push({ task, dueDate });
      }
    } catch (error) {
      // Skip tasks with invalid date format
      continue;
    }
  }
  
  // Sort by urgency (nearest due first)
  if (upcomingTasks.length > 0) {
    upcomingTasks.sort((a, b) => a.dueDate.getTime() - b.dueDate.getTime());
    const { task: mostUrgentTask, dueDate } = upcomingTasks[0];
    
    // Calculate time until due
    const timeUntilDue = dueDate.getTime() - now.getTime();
    const hoursUntil = timeUntilDue / (1000 * 60 * 60);
    
    // Format message based on urgency
    let urgency: "CRITICAL" | "URGENT" | "HIGH" | "NORMAL";
    if (hoursUntil < 2) {
      urgency = "CRITICAL";
    } else if (hoursUntil < 24) {
      urgency = "URGENT";
    } else if (hoursUntil < 48) {
      urgency = "HIGH";
    } else {
      urgency = "NORMAL";
    }
    
    let timeStr: string;
    if (hoursUntil < 1) {
      timeStr = `${Math.floor(timeUntilDue / (1000 * 60))} minutes`;
    } else if (hoursUntil < 24) {
      timeStr = `${Math.floor(hoursUntil)} hours`;
    } else {
      const days = Math.floor(hoursUntil / 24);
      timeStr = `${days} days`;
    }
    
    const message = `Focus on ${mostUrgentTask.course} - ${mostUrgentTask.title} | Due in ${timeStr} (${urgency})`;
    
    return {
      type: "TASK",
      message,
      data: {
        task: mostUrgentTask,
        hoursUntil: Math.round(hoursUntil * 100) / 100,
        urgency,
      },
    };
  }
  
  // No urgent tasks - check if class starting soon (within 60 min)
  const allEvents = await getUserSchedule(userId);
  
  // Get current day of week
  const currentDay = now.toLocaleDateString('en-US', { weekday: 'long' });
  const currentTimeStr = `${now.getHours().toString().padStart(2, '0')}${now.getMinutes().toString().padStart(2, '0')}`;
  const currentTimeInt = parseInt(currentTimeStr, 10);
  
  // Check events for today or next few days
  const dayOrder = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
  const currentDayIdx = dayOrder.indexOf(currentDay) >= 0 ? dayOrder.indexOf(currentDay) : 0;
  
  for (const event of allEvents) {
    const eventDayIdx = dayOrder.indexOf(event.day) >= 0 ? dayOrder.indexOf(event.day) : 99;
    
    // Check if event is today or in the future
    const daysUntilEvent = (eventDayIdx - currentDayIdx + 7) % 7;
    
    if (daysUntilEvent === 0) {
      // Today
      try {
        const eventStartInt = parseInt(event.startTime, 10);
        const timeDiff = eventStartInt - currentTimeInt;
        // Convert to minutes
        const hours = Math.floor(timeDiff / 100);
        const minutes = timeDiff % 100;
        const minutesUntil = hours * 60 + minutes;
        
        if (minutesUntil >= 0 && minutesUntil <= 60) {
          const startTimeFormatted = event.startTime.length === 4 
            ? `${event.startTime.slice(0, 2)}:${event.startTime.slice(2)}` 
            : event.startTime;
          const message = `You have ${event.module} ${event.type} starting at ${startTimeFormatted} in ${minutesUntil} minutes`;
          return {
            type: "CLASS",
            message,
            data: {
              event,
              minutesUntil,
            },
          };
        }
      } catch (error) {
        continue;
      }
    } else if (daysUntilEvent > 0 && daysUntilEvent <= 1) {
      // Tomorrow or next day
      try {
        const eventStartInt = parseInt(event.startTime, 10);
        if (eventStartInt <= 1400) {
          // Before 2 PM
          const startTimeFormatted = event.startTime.length === 4 
            ? `${event.startTime.slice(0, 2)}:${event.startTime.slice(2)}` 
            : event.startTime;
          const message = `You have ${event.module} ${event.type} tomorrow at ${startTimeFormatted}`;
          return {
            type: "CLASS",
            message,
            data: {
              event,
              daysUntil: daysUntilEvent,
            },
          };
        }
      } catch (error) {
        continue;
      }
    }
  }
  
  // No urgent tasks or upcoming classes
  return {
    type: "IDLE",
    message: "No urgent tasks. You can work ahead!",
  };
}
